### PR TITLE
Static parsing

### PR DESCRIPTION
**Pull request by BugDiver**
Added support for Gauge LSP Diagonastics, Go to Definition, Code Action

### CODE REVIEW

File: gauge-ruby.go [hunk 1]
```diff
@@ -163,7 +163,7 @@ func main() {
 	}
 	if *start {
 		os.Chdir(projectRoot)
-		err = runCommand("ruby", "-e", "require 'gauge_runtime'")
+		err = runCommand("bundle", "exec", "ruby", "-e", "require 'gauge_runtime'")
```
Comment (nehashri) [inline gauge-ruby.go]: Why has this changed?

File: lib/connector.rb [hunk 2]
```diff
@@ -37,37 +37,14 @@ def self.executionSocket
...
-      apiMessage.serialize(@@apiSocket)
-
-      responseLen = message_length(@@apiSocket)
-      data = @@apiSocket.read responseLen
-      message = Gauge::Messages::APIMessage.parse(data)
-    end
-
-    def self.get_unique_id
-      rand(2**63-1)
+      return text.gsub(/(<.*?>)/, "{}")
```
Comment (nehashri) [inline lib/connector.rb]: Any reason for this change?

File: lib/processors/execution_hook_processors.rb [hunk 3]
```diff
@@ -17,12 +17,15 @@
...
-
+require_relative "../executor"
+require_relative "../method_cache"
 module Gauge
   module Processors
     include ExecutionHandler
 
     def process_execution_start_request(message)
+      Gauge::MethodCache.clear
+      Executor.load_steps(File.join(ENV["GAUGE_PROJECT_ROOT"], 'step_implementations'))
```
Comment (nehashri) [inline lib/processors/execution_hook_processors.rb]: The method step implementations path can be moved to a common place

File: lib/gauge_runtime.rb [hunk 4]
```diff
@@ -20,14 +20,15 @@
...
+require_relative 'static_loader'
 require_relative 'connector'
 require_relative 'message_processor'
 
 
 module Gauge
   # @api private
   module Runtime
-    DEFAULT_IMPLEMENTATIONS_DIR_PATH = File.join(Dir.pwd, 'step_implementations')
+    DEFAULT_IMPLEMENTATIONS_DIR_PATH = File.join(ENV["GAUGE_PROJECT_ROOT"], 'step_implementations')
```
Comment (nehashri) [inline lib/gauge_runtime.rb]: The method step implementations path can be moved to a common place

File: lib/processors/cache_file_processor.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        ast = CodeParser.code_to_ast(message.cacheFileRequest.content)
+        StaticLoader.reload_steps(message.cacheFileRequest.filePath, ast)
+      else
+        load_from_disk message.cacheFileRequest.filePath
+      end
+      nil
+    end
+
+    def load_from_disk(f)
+      ast = CodeParser.code_to_ast File.read(f)
```
Comment (nehashri) [inline lib/processors/cache_file_processor.rb]: need to create ast only if file exists.

File: lib/connector.rb (hunk 2 above)
Comment (BugDiver) [reply lib/connector.rb]: We used to make an API call from runner to gauge to get stepValue of a stepText. for that, we end up having two socket connections.  As per discussion with @sriv  and @Apoorva-GA we found that there was no specific reason for that. It was just a legacy which came form gauge-java.  So we decided to remove it.

File: gauge-ruby.go (hunk 1 above)
Comment (BugDiver) [reply gauge-ruby.go]: When start runner for execution from command line we use `bundle exec gauge run` which make sure all your required gems as available for the project. Whiling starting the runner for LSP we need to do the same or else `gauge-ruby` gem will not be available. or users have to install `gauge-ruby` globally.

**Pull request merged**